### PR TITLE
The PTMLBuilder 

### PR DESCRIPTION
**Pull request by dwarfyresearch**

### CODE REVIEW

File: include/revng/PTML/Tag.h [hunk 1]
```diff
@@ -142,4 +160,24 @@ inline ScopeTag Tag::scope(llvm::raw_ostream &OS, bool Newline) const {
...
+public:
+  PTMLBuilder(bool GenerateTagLessPTML = false);
+
+public:
+  bool isGenerateTagLessPTML() const;
+
+public:
+  ptml::Tag getTag(llvm::StringRef Tag) const;
+  ptml::Tag getTag(llvm::StringRef Tag, llvm::StringRef Content) const;
+  ptml::Tag getTaglessTag(llvm::StringRef Content = llvm::StringRef()) const;
```
[the remark this answers is not in the document]
Comment (dwarfyresearch) [reply include/revng/PTML/Tag.h:176]: I think yes. It is kinda leftover. Thanks.

File: lib/PTML/Tag.cpp [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+PTMLBuilder::PTMLBuilder(bool GenerateTagLessPTML) :
+  GenerateTagLessPTML(GenerateTagLessPTML) {
+}
+
+bool PTMLBuilder::isGenerateTagLessPTML() const {
+  return GenerateTagLessPTML;
+}
+
+ptml::Tag PTMLBuilder::getTag(llvm::StringRef Tag) const {
+  return ptml::Tag(Tag);
```
Comment (pfez) [inline lib/PTML/Tag.cpp:19]: This should check if the `PTMLBuilder` is tagless and emit an empty tag if that's the case.

File: lib/PTML/Tag.cpp [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+ptml::Tag
+PTMLBuilder::getTag(llvm::StringRef Tag, llvm::StringRef Content) const {
+  return ptml::Tag(Tag, Content);
```
Comment (pfez) [inline lib/PTML/Tag.cpp:24]: This should check if the `PTMLBuilder` is tagless and emit an empty tag if that's the case.

File: lib/PTML/Tag.cpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+ptml::Tag PTMLBuilder::getTaglessTag(llvm::StringRef Content) const {
+  revng_assert(GenerateTagLessPTML);
+  Tag EmptyTag;
+  EmptyTag.setContent(Content);
+  return EmptyTag;
+}
```
Comment (pfez) [inline lib/PTML/Tag.cpp:32]: This should go away.

File: lib/PTML/IndentedOstream.cpp [hunk 5]
```diff
@@ -40,9 +40,15 @@ void PTMLIndentedOstream::write_impl(const char *Ptr, size_t Size) {
...
-    OS << Tag(tags::Span, std::string(IndentSize * IndentDepth, ' '))
-            .addAttribute(attributes::Token, ptml::tokens::Indentation)
-            .serialize();
+    if (!ThePTMLBuilder.isGenerateTagLessPTML()) {
+      OS << ThePTMLBuilder
+              .getTag(tags::Span, std::string(IndentSize * IndentDepth, ' '))
+              .addAttribute(attributes::Token, ptml::tokens::Indentation)
+              .serialize();
+    } else {
+      // TODO: Check if we need this `else`.
```
[the remark this answers is not in the document]
Comment (pfez) [reply lib/PTML/IndentedOstream.cpp:49]: Ah wait. If `getTag` is edited as I proposed above, to emit tagless tags on demand if necessary, then the `if` is not even necessary anymore. You can just leave the `writeIndent` method as it was before.

[the remark this answers is not in the document]
Comment (dwarfyresearch) [reply lib/PTML/IndentedOstream.cpp:49]: Yes, right

**Comment by pfez** [comment]
Merged in f05c08547dd18f4219fa167abf60787e670f304d

Thanks!

**Pull request closed without merging**